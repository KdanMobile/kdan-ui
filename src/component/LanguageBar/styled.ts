import styled, { css, CSSObject } from "styled-components";
import { colors } from "../../themes";

type Theme = "default" | "dark";

type StyleType = {
  borderTop: string;
  color: string;
  backgroundColor: string;
  hover: CSSObject;
};

type ThemeHubType = {
  default: StyleType;
  dark: StyleType;
};

const themeHub: ThemeHubType = {
  default: {
    borderTop: `5px solid ${colors.N100}`,
    color: "rgb(64, 64, 64)",
    backgroundColor: colors.N0,
    hover: {
      color: "rgb(64, 64, 64)",
      backgroundColor: "#e6e6e6",
    },
  },
  dark: {
    borderTop: `5px solid ${colors.N0}`,
    color: colors.N0,
    backgroundColor: colors.N85,
    hover: {
      color: colors.N85,
      backgroundColor: "rgba(0, 0, 0, 0.1)",
    },
  },
};

export const BarWrapper = styled.div`
  position: relative;
  display: block;
  margin: 15px auto 0;
  user-select: none;
  font-size: 0.8125rem;
`;

const BlockStyle = ({ isItem, theme }: { isItem: boolean; theme: Theme }) => {
  if (isItem) {
    return css`
      &:first-child {
        border-top-left-radius: 4px;
        border-top-right-radius: 4px;
      }
      &:last-child {
        border-bottom-left-radius: 4px;
        border-bottom-right-radius: 4px;
      }
      &:only-child {
        border-radius: 4px;
      }
      &:hover {
        color: ${themeHub[theme].hover.color};
        background-color: ${themeHub[theme].hover.backgroundColor};
      }
    `;
  } else {
    return css`
      border-radius: 4px;
    `;
  }
};

export const Arrow = styled.span`
  width: 0;
  height: 0;
  border-left: 5px solid transparent;
  border-right: 5px solid transparent;
  margin: auto 0;

  ${({ isOpen }: { isOpen: boolean }) =>
    isOpen &&
    css`
      -moz-transform: rotate(-180deg);
      -webkit-transform: rotate(-180deg);
      -o-transform: rotate(-180deg);
      -ms-transform: rotate(-180deg);
      transform: rotate(-180deg);
    `};
`;

export const BlockWrapper = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
  width: 120px;
  padding: 5px 10px;
  margin: auto;
  font-size: 14px;
  cursor: pointer;

  color: ${({ theme }: { theme: Theme }) => themeHub[theme].color};
  background-color: ${({ theme }: { theme: Theme }) =>
    themeHub[theme].backgroundColor};

  ${BlockStyle};

  ${Arrow} {
    border-top: ${({ theme }: { theme: Theme }) => themeHub[theme].borderTop};
  }
`;

export const Menu = styled.div<{ isOpen: boolean }>`
  position: absolute;
  display: ${({ isOpen }) => (isOpen ? "block" : "none")};
  bottom: calc(100% + 3px);
  left: 50%;
  transform: translateX(-50%);
  background: ${colors.N0};
  border-radius: 4px;
  box-shadow: 0 2px 4px 0 rgba(0, 0, 0, 0.1);
`;
