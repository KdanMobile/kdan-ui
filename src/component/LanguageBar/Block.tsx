import React from 'react';
import PropTypes from 'prop-types';

import { BlockWrapper, Arrow } from './styled';

type Props = {
  value: React.ReactNode;
  isItem?: boolean;
  onChange?: () => void;
  isOpen?: boolean;
  theme?: string;
};

const Block: React.FC<Props> = ({
  value,
  isItem = false,
  onChange = () => {},
  isOpen = false,
  theme,
}) => {
  return (
    <BlockWrapper isItem={isItem} onClick={onChange} theme={theme}>
      {value}
      {!isItem && <Arrow isOpen={isOpen} />}
    </BlockWrapper>
  );
};

export default Block;
