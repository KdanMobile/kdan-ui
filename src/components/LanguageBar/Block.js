import React from 'react';
import PropTypes from 'prop-types';

import { BlockWrapper, Arrow } from './styled';

const Block = ({ value, isItem, onChange, isOpen, theme }) => {
  return (
    <BlockWrapper isItem={isItem} onClick={onChange} theme={theme}>
      {value}
      {!isItem && <Arrow isOpen={isOpen} />}
    </BlockWrapper>
  );
};

Block.defaultProps = {
  isItem: false,
  onChange: () => {},
  isOpen: false
};

Block.prototype = {
  value: PropTypes.string.isRequired,
  isItem: PropTypes.bool,
  onChange: PropTypes.func,
  isOpen: PropTypes.bool
};

export default Block;
