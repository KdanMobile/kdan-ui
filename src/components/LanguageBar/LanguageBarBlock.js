import React from 'react';
import PropTypes from 'prop-types';

import { BlockWrapper, BarIcon } from './styled';

const LanguageBarBlock = ({ value, isItem, onChange, ...otherProps }) => {
  return (
    <BlockWrapper isItem={isItem} onClick={onChange} {...otherProps}>
      {value}
      {!isItem && <BarIcon />}
    </BlockWrapper>
  );
};

LanguageBarBlock.defaultProps = {
  isItem: false,
  onChange: () => {}
};

LanguageBarBlock.prototype = {
  value: PropTypes.string.isRequired,
  isItem: PropTypes.bool,
  onChange: PropTypes.func
};

export default LanguageBarBlock;
