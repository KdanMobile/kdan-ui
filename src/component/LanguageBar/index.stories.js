import React from 'react';
import { action } from '@storybook/addon-actions';

import LanguageBar from './';
import { Container } from '../Decorator';

const langLabels = [
  {
    key: 'en',
    value: 'English'
  },
  {
    key: 'ja',
    value: '日本語'
  },
  {
    key: 'zh-tw',
    value: '中文（繁體）'
  },
  {
    key: 'zh-cn',
    value: '中文（简体）'
  },
  {
    key: 'es',
    value: 'Español'
  },
  {
    key: 'ko',
    value: '한국어'
  },
  {
    key: 'it',
    value: 'Italiano'
  }
];

export default {
  component: LanguageBar,
  title: 'LanguageBar',
  decorators: [(story) => <Container size='middle'>{story()}</Container>],
  parameters: {
    docs: {
      description: {
        component: '下拉式語系選單。'
      }
    }
  }
};

const actionsData = {
  onChange: action('onChange')
};

export const Base = (args) => <LanguageBar {...args} />;
Base.args = {
  theme: 'default',
  value: 'English',
  options: langLabels,
  ...actionsData
};
