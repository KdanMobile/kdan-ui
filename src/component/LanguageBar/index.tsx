import React, { useState } from "react";

import { BarWrapper, Menu } from "./styled";

import Block from "./Block";
import ClickAwayListener from "../ClickAwayListener";

type OptionType = {
  key: string | number;
  value: React.ReactNode;
};

type Props = {
  options: OptionType[];
  value: React.ReactNode;
  onChange?: (value: OptionType) => void;
  theme?: "default" | "dark";
};

const LanguageBar: React.FC<Props> = ({
  options,
  value: defaultValue,
  onChange = () => {
    // do something;
  },
  theme = "default",
}: Props) => {
  const [isOpen, setIsOpen] = useState(false);
  const [value, setValue] = useState(defaultValue);

  const onChangeHandle = (option: OptionType) => {
    onChange(option);
    setValue(option.value);
  };

  const handleClickAway = () => {
    if (isOpen) {
      setIsOpen(false);
    }
  };

  return (
    <ClickAwayListener onClick={handleClickAway}>
      <BarWrapper onClick={() => setIsOpen(!isOpen)} data-testid="languageBar">
        <Menu isOpen={isOpen}>
          {options &&
            options.map((option, index) => (
              <Block
                theme={theme}
                isItem
                key={`${index}-${option.key}`}
                value={option.value}
                onChange={() => onChangeHandle(option)}
              />
            ))}
        </Menu>
        <Block value={value} theme={theme} isOpen={isOpen} />
      </BarWrapper>
    </ClickAwayListener>
  );
};

export default LanguageBar;
