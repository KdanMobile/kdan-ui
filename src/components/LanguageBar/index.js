import React, { useState } from 'react';
import PropTypes from 'prop-types';

import { BarWrapper, Menu } from './styled';

import Block from './Block';
import ClickAwayListener from '../ClickAwayListener';

const LanguageBar = ({ options, value: defaultValue, onChange, theme }) => {
  const [isOpen, setIsOpen] = useState(false);
  const [value, setValue] = useState(defaultValue);

  const onChangeHandle = (option) => {
    onChange(option);
    setValue(option.value);
  };

  const handleClickAway = () => {
    if (isOpen) {
      setIsOpen(false);
    }
  };

  return (
    <ClickAwayListener onClick={handleClickAway}>
      <BarWrapper onClick={() => setIsOpen(!isOpen)}>
        <Menu isOpen={isOpen}>
          {options.map((option, index) => (
            <Block
              theme={theme}
              isItem
              key={`${index}-${option.key}`}
              value={option.value}
              onChange={() => onChangeHandle(option)}
            />
          ))}
        </Menu>
        <Block value={value} theme={theme} isOpen={isOpen} />
      </BarWrapper>
    </ClickAwayListener>
  );
};

LanguageBar.defaultProps = {
  onChange: () => {},
  theme: 'default'
};

LanguageBar.propTypes = {
  options: PropTypes.array.isRequired,
  /**
    select 目前的值
  */
  value: PropTypes.string.isRequired,
  /**
    選擇後觸發事件，會回傳所選那個物件。
  */
  onChange: PropTypes.func,
  theme: PropTypes.oneOf(['default', 'dark'])
};

export default LanguageBar;
