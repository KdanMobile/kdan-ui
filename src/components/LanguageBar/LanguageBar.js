import React, { useState } from 'react';
import PropTypes from 'prop-types';

import { BarWrapper, Menu } from './styled';

import LanguageBarBlock from './LanguageBarBlock';
import { ClickAwayListener } from '../ClickAwayListener';

const LanguageBar = ({ options, value: defaultValue, onChange, theme }) => {
  const [isOpen, setIsOpen] = useState(false);
  const [value, setValue] = useState(defaultValue);

  const onChangeHandle = (option) => {
    onChange(option);
    setValue(option.value);
  };

  const handleClickAway = () => {
    if (isOpen) {
      setIsOpen(false);
    }
  };

  return (
    <ClickAwayListener onClick={handleClickAway}>
      <BarWrapper onClick={() => setIsOpen(!isOpen)}>
        <Menu isOpen={isOpen}>
          {options.map((option, index) => (
            <LanguageBarBlock
              theme={theme}
              isItem
              key={`${index}-${option.key}`}
              value={option.value}
              onChange={() => onChangeHandle(option)}
            />
          ))}
        </Menu>
        <LanguageBarBlock value={value} theme={theme} isOpen={isOpen} />
      </BarWrapper>
    </ClickAwayListener>
  );
};

LanguageBar.defaultProps = {
  onChange: () => {},
  theme: 'default'
};

LanguageBar.propTypes = {
  options: PropTypes.array.isRequired,
  value: PropTypes.string.isRequired,
  onChange: PropTypes.func,
  theme: PropTypes.oneOf(['default', 'dark'])
};

export { LanguageBar };
