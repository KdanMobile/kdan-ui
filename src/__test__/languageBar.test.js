import React from 'react';
import { render, fireEvent, cleanup } from '@testing-library/react';
import 'jest-styled-components';
import '@testing-library/jest-dom';

import LanguageBar from '../components/LanguageBar';

describe('LanguageBar', () => {
  afterEach(cleanup);

  const options = [
    {
      key: 'zh-tw',
      value: '中文'
    },
    {
      key: 'en',
      value: '英文'
    }
  ];
  const value = 'zh-tw';

  test('basic', () => {
    const { container } = render(
      <LanguageBar options={options} value={value} />
    );
    expect(container).toMatchSnapshot();
  });

  test('should show menu on click', () => {
    const { container } = render(
      <LanguageBar options={options} value={value} />
    );
    fireEvent.click(container.querySelector('.sc-AxjAm'));
    expect(container.querySelector('.sc-AxhCb')).toHaveStyleRule(
      'display',
      'block'
    );
  });

  test('should change value on click', () => {
    const { container } = render(
      <LanguageBar options={options} value={value} />
    );
    fireEvent.click(container.querySelectorAll('.sc-AxiKw')[1]);
    expect(container.querySelector('.jXPJcR').textContent).toBe('英文');
  });
});
