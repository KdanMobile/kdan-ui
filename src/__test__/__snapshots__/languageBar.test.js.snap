// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LanguageBar basic 1`] = `
.c0 {
  position: relative;
  display: block;
  margin: 15px auto 0;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  font-size: 0.8125rem;
}

.c5 {
  width: 0;
  height: 0;
  border-left: 5px solid transparent;
  border-right: 5px solid transparent;
  margin: auto 0;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 120px;
  padding: 5px 10px;
  margin: auto;
  font-size: 14px;
  cursor: pointer;
  color: rgb(64,64,64);
  background-color: #ffffff;
}

.c2:first-child {
  border-top-left-radius: 4px;
  border-top-right-radius: 4px;
}

.c2:last-child {
  border-bottom-left-radius: 4px;
  border-bottom-right-radius: 4px;
}

.c2:only-child {
  border-radius: 4px;
}

.c2:hover {
  color: rgb(64,64,64);
  background-color: #e6e6e6;
}

.c2 .c4 {
  border-top: 5px solid #000000;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 120px;
  padding: 5px 10px;
  margin: auto;
  font-size: 14px;
  cursor: pointer;
  color: rgb(64,64,64);
  background-color: #ffffff;
  border-radius: 4px;
}

.c3 .c4 {
  border-top: 5px solid #000000;
}

.c1 {
  position: absolute;
  display: none;
  bottom: calc(100% + 3px);
  left: 50%;
  -webkit-transform: translateX(-50%);
  -ms-transform: translateX(-50%);
  transform: translateX(-50%);
  background: #ffffff;
  border-radius: 4px;
  box-shadow: 0 2px 4px 0 rgba(0,0,0,0.1);
}

<div>
  <div>
    <div
      class="c0"
    >
      <div
        class="c1"
      >
        <div
          class="c2"
        >
          中文
        </div>
        <div
          class="c2"
        >
          英文
        </div>
      </div>
      <div
        class="c3"
      >
        zh-tw
        <span
          class="c4 c5"
        />
      </div>
    </div>
  </div>
</div>
`;
